static build of postgresql (and thus libpq) fails

### Steps To Reproduce
Steps to reproduce the behavior:
1. patch nixpkgs to fix lz4 static build (https://github.com/NixOS/nixpkgs/pull/191915)
2. build statically some postgresql client tool that links libpq, e.g. pg_top
E.g., `nix build github:robx/nixpkgs/lz4-static#pkgsStatic.pg_top` does the job.

### Build log
```
$ nix build github:robx/nixpkgs/lz4-static#pkgsStatic.pg_top
error: builder for '/nix/store/jc4h3yc4k5n478bvchfccdyj1a8g3fbx-postgresql-static-x86_64-unknown-linux-musl-14.5.drv' failed with exit code 2;
       last 10 log lines:
       > /nix/store/q8zpwwa77zaa5h0p15iiid26g5ak6s7l-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: /nix/store/zbyfyc4a71mymphik6kz8qxsfasaf1mc-icu4c-71.1-x86_64-unknown-linux-musl/lib/libicuuc.a(ustack.ao):(.data.rel.ro._ZTIN6icu_716UStackE[_ZTIN6icu_716UStackE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
       > /nix/store/q8zpwwa77zaa5h0p15iiid26g5ak6s7l-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: /nix/store/zbyfyc4a71mymphik6kz8qxsfasaf1mc-icu4c-71.1-x86_64-unknown-linux-musl/lib/libicuuc.a(chariter.ao):(.data.rel.ro._ZTIN6icu_7124ForwardCharacterIteratorE[_ZTIN6icu_7124ForwardCharacterIteratorE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
       > /nix/store/q8zpwwa77zaa5h0p15iiid26g5ak6s7l-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: /nix/store/zbyfyc4a71mymphik6kz8qxsfasaf1mc-icu4c-71.1-x86_64-unknown-linux-musl/lib/libicuuc.a(chariter.ao):(.data.rel.ro._ZTIN6icu_7117CharacterIteratorE[_ZTIN6icu_7117CharacterIteratorE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
       > /nix/store/q8zpwwa77zaa5h0p15iiid26g5ak6s7l-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: /nix/store/zbyfyc4a71mymphik6kz8qxsfasaf1mc-icu4c-71.1-x86_64-unknown-linux-musl/lib/libicuuc.a(parsepos.ao):(.data.rel.ro._ZTIN6icu_7113ParsePositionE[_ZTIN6icu_7113ParsePositionE]+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
       > collect2: error: ld returned 1 exit status
       > make[2]: *** [Makefile:66: postgres] Error 1
       > make[2]: Leaving directory '/build/postgresql-14.5/src/backend'
       > make[1]: *** [Makefile:42: all-backend-recurse] Error 2
       > make[1]: Leaving directory '/build/postgresql-14.5/src'
       > make: *** [GNUmakefile:16: world-src-recurse] Error 2
       For full logs, run 'nix log /nix/store/jc4h3yc4k5n478bvchfccdyj1a8g3fbx-postgresql-static-x86_64-unknown-linux-musl-14.5.drv'.
error: 1 dependencies of derivation '/nix/store/lxlnl2kal1r6a7s6rihs51809djrs79m-pg_top-static-x86_64-unknown-linux-musl-3.7.0.drv' failed to build
```

### Additional context

For the purpose of statically building tools that use `libpq`, we wouldn't actually need a static build of the postgresql backend. Compare https://github.com/NixOS/nixpkgs/issues/61580, which suggests a separate `libpq` package.

### Notify maintainers
@thoughtpolice @marsam
 
### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix-shell -p nix-info --run "nix-info -m"
this path will be fetched (0.06 MiB download, 0.30 MiB unpacked):
  /nix/store/q6ayjwizl0xcbvknkcksy5g4n1ipa6b1-bash-interactive-5.1-p8-dev
copying path '/nix/store/q6ayjwizl0xcbvknkcksy5g4n1ipa6b1-bash-interactive-5.1-p8-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.91, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Setting `icuEnabled` to false, the build fails with

```
/nix/store/q8zpwwa77zaa5h0p15iiid26g5ak6s7l-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld:
  /nix/store/bgsjggllvcb8bs47mnpyz25rq0jvb7qz-libxml2-static-x86_64-unknown-linux-musl-2.9.14/lib/libxml2.a(nanohttp.o):
    in function `xmlNanoHTTPRead':
      > (.text+0x10f8): undefined reference to `inflate'
```

Some progress:
- xmlstarlet is independently broken (e.g. it should be calling `$PKG_CONFIG` instead of `pkg-config`, though that's not enough to fix things)
- `$PKG_CONFIG` output seems correct (i.e., with `--libs libxml-2.0 --static`, it includes proper flags to link zlib)
- the zlib linking error goes away by patching this buggy `Makefile` line to not filter it: https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/backend/Makefile;h=0da848b1fd3dafeebf4c9fe0d4ff3e5e2c8c5e22;hb=278273ccbad27a8834dfdf11895da9cd91de4114#l51 (that's in postgresql 14.5, since fixed)
- now there's a new failure:
```
make[3]: Entering directory '/home/rob/postgresql-14.5/src/backend/utils/mb/conversion_procs/euc_cn_and_mic'
x86_64-unknown-linux-musl-cc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -O2 -fPIC -I../../../../../../src/include  -D_GNU_SOURCE -I/nix/store/cks2a0805w2scxjavn8299fawadng2i6-libxml2-static-x86_64-unknown-linux-musl-2.9.14-dev/include/libxml2 -I/nix/store/yy6c26ms94fff2nckjrfjm416f7k5hvb-lz4-static-x86_64-unknown-linux-musl-1.9.4-dev/include   -c -o euc_cn_and_mic.o euc_cn_and_mic.c
x86_64-unknown-linux-musl-cc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -O2 -fPIC -shared -o euc_cn_and_mic.so euc_cn_and_mic.o  -L../../../../../../src/port -L../../../../../../src/common   -L/nix/store/bgsjggllvcb8bs47mnpyz25rq0jvb7qz-libxml2-static-x86_64-unknown-linux-musl-2.9.14/lib -L/nix/store/6hj73llfmli3d3c67lmhwc7490hx24wj-zlib-static-x86_64-unknown-linux-musl-1.2.12/lib -L/nix/store/qqq58sk2qgicygjk7z91m99kns87xm3g-lz4-static-x86_64-unknown-linux-musl-1.9.4/lib    
/nix/store/q8zpwwa77zaa5h0p15iiid26g5ak6s7l-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: /nix/store/ravnd17fg4maxzc5ns8vbijn7ah3p4b1-x86_64-unknown-linux-musl-stage-final-gcc-11.3.0/lib/gcc/x86_64-unknown-linux-musl/11.3.0/crtbeginT.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a shared object
/nix/store/q8zpwwa77zaa5h0p15iiid26g5ak6s7l-x86_64-unknown-linux-musl-binutils-2.38/bin/x86_64-unknown-linux-musl-ld: failed to set dynamic section sizes: bad value
collect2: error: ld returned 1 exit status
```

--- a/pkgs/servers/sql/postgresql/generic.nix
+++ b/pkgs/servers/sql/postgresql/generic.nix
@@ -53,7 +53,11 @@ let
       , libkrb5
 
       # icu
-      , icuSupport ? true
+      # Building with icu in pkgsStatic gives tons of "undefined reference" errors like this:
+      #   /nix/store/452lkaak37d3mzzn3p9ak7aa3wzhdqaj-icu4c-74.2-x86_64-unknown-linux-musl/lib/libicuuc.a(chariter.ao):
+      #    (.data.rel.ro._ZTIN6icu_7417CharacterIteratorE[_ZTIN6icu_7417CharacterIteratorE]+0x0):
+      #    undefined reference to `vtable for __cxxabiv1::__si_class_type_info'
+      , icuSupport ? !stdenv.hostPlatform.isStatic
       , icu
 
       # JIT
@@ -71,7 +75,11 @@ let
       , gettext
 
       # PAM
-      , pamSupport ? stdenv.hostPlatform.isLinux
+      # Building with linux-pam in pkgsStatic gives a few "undefined reference" errors like this:
+      #   /nix/store/3s55icpsbc36sgn7sa8q3qq4z6al6rlr-linux-pam-static-x86_64-unknown-linux-musl-1.6.1/lib/libpam.a(pam_audit.o):
+      #     in function `pam_modutil_audit_write':(.text+0x571):
+      #     undefined reference to `audit_close'
+      , pamSupport ? stdenv.hostPlatform.isLinux && !stdenv.hostPlatform.isStatic
       , linux-pam
 
       # PL/Perl
@@ -267,7 +275,7 @@ let
         # because there is a realistic use-case for extensions to locate the /lib directory to
         # load other shared modules.
         remove-references-to -t "$dev" -t "$doc" -t "$man" "$out/bin/postgres"
-
+      '' + lib.optionalString (!stdenv'.hostPlatform.isStatic) ''
         if [ -z "''${dontDisableStatic:-}" ]; then
           # Remove static libraries in case dynamic are available.
           for i in $lib/lib/*.a; do
@@ -278,6 +286,7 @@ let
             fi
           done
         fi
+      '' + ''
         # The remaining static libraries are libpgcommon.a, libpgport.a and related.
         # Those are only used when building e.g. extensions, so go to $dev.
         moveToOutput "lib/*.a" "$dev"
@@ -306,7 +315,7 @@ let
     # Also see <nixpkgs>/doc/stdenv/platform-notes.chapter.md
     doCheck = false;
     # Tests just get stuck on macOS 14.x for v13 and v14
-    doInstallCheck = !(stdenv'.hostPlatform.isDarwin && olderThan "15");
+    doInstallCheck = !(stdenv'.hostPlatform.isDarwin && olderThan "15") && !(stdenv'.hostPlatform.isStatic);
     installCheckTarget = "check-world";
 
     passthru = let

--- a/pkgs/servers/sql/postgresql/generic.nix
+++ b/pkgs/servers/sql/postgresql/generic.nix
@@ -149,6 +149,8 @@ let
       ];
     };
 
+    strictDeps = true;
+
     buildInputs = [
       zlib
       readline
@@ -171,6 +173,7 @@ let
       ++ lib.optionals nlsSupport [ gettext ];
 
     nativeBuildInputs = [
+      libxml2
       makeWrapper
       pkg-config
       removeReferencesTo

--- a/pkgs/servers/sql/postgresql/generic.nix
+++ b/pkgs/servers/sql/postgresql/generic.nix
@@ -218,9 +218,8 @@ let
       ++ lib.optionals pythonSupport [ "--with-python" ]
       ++ lib.optionals jitSupport [ "--with-llvm" ]
       ++ lib.optionals pamSupport [ "--with-pam" ]
-      # This could be removed once the upstream issue is resolved:
-      # https://postgr.es/m/flat/427c7c25-e8e1-4fc5-a1fb-01ceff185e5b%40technowledgy.de
-      ++ lib.optionals (stdenv'.hostPlatform.isDarwin && atLeast "16") [ "LDFLAGS_EX_BE=-Wl,-export_dynamic" ]
+      # This can be removed once v17 is removed. v18+ ships with it.
+      ++ lib.optionals (stdenv'.hostPlatform.isDarwin && atLeast "16" && olderThan "18") [ "LDFLAGS_EX_BE=-Wl,-export_dynamic" ]
       ++ lib.optionals (atLeast "17" && !perlSupport) [ "--without-perl" ]
       ++ lib.optionals ldapSupport [ "--with-ldap" ]
       ++ lib.optionals tclSupport [ "--with-tcl" ]

